lib/init.php

```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
 	);
 }
 add_action( 'admin_menu', 'gutenberg_menu', 9 );
+
+if ( ! function_exists( 'html_contains_block' ) ) {
+	/**
+	 * Recursively find a block by attribute.
+	 *
+	 * @param  string  $html The html to search in.
+	 * @param  string  $block_name The block type to search for.
+	 * @param  string  $attribute_name The attribute name to search for.
+	 * @param  string  $attribute_value The attribute value to search for.
+	 *
+	 * @return bool    True if block is found, false otherwise
+	 */
+	function html_contains_block( $html, $block_name, $attribute_name = null, $attribute_value = null ) {
+		$at = 0;
+
+		/**
+		 * This is the same regex as the one used in the block parser.
+		 * It is better to use this solution to look for a block's existence
+		 * in a document compared to having to parsing the blocks in the
+		 * document, avoiding all the performance drawbacks of achieving
+		 * a full representation of block content just to check if one block
+		 * is there.
+		 */
+		$pattern = sprintf(
+			'~<!--\s+?wp:%s\s+(?P<attrs>{(?:(?:[^}]+|}+(?=})|(?!}\s+/?-->).)*+)?}\s+)?/?-->~s',
+			// @TODO: we could ensure that there's a namespace on the block name.
+			preg_quote( $block_name, '~' )
```

Comment:
this is more or less a decision to make, whether to support the optional core namespace, but it's easy enough to augment.

```php
$full_block_name = str_contains( $block_name, '/' ) ? $block_name : "core/{$block_name}";
```

Reply:
I'd prefer if we weren't clever about this and just required the block name in full. I don't want to have to look at the code to find out why passing just `navigation` magically works 🙏

Reply:
this is a fair take, though it can be balanced against the fact that the use of an implicit `core/` namespace permeates Gutenberg, so the choice not to do it here is the surprising behavior. personally I don't care much as long as it's clear in the function documentation, but it's not particularly "clever" to support a well-spec'd and near-universal behavior.

Reply:
yeah @draganescu something happened with the inline code spans in your response, which unfortunately made it hard for me to clearly interpret.

I'll rephrase what I think you might have already denoted:

 - a search for `paragraph` should match on `paragraph` or `core/paragraph`.
 - `<!-- wp:paragraph -->` should match for `paragraph` and `core/paragraph`.
 - a search for `paragraph` _should not_ match on `kadence/paragraph`.
 - `<!-- wp:paragraph -->` _should not_ match for `kadence/paragraph`.

So basically we have to treat all block names without a `core/` namespace as if they had one.

```php
$block_type = str_contains( $block_name, '/' ) ? $block_name : "core/{$block_name}";
preg_match( '... wp:(P<namespace> ... /)?(P<name> ... ) ...', ... );
$found_block_type = empty( $match['namespace'] ) ? "core/{$match['name']}" : $match['name'];

if ( $block_type === $found_block_type ) {
	…
}
```